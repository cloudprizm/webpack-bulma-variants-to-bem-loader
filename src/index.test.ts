import * as Path from 'path'
import * as Fs from 'fs'
import {
  convertToUnionType,
  getAllStylesFromFile,
  getBlockNameFromPath,
  combineExistingUnionsToAll,
  getExportKeys,
  exportBEMTypes,
  bulmaToBEM
} from './index'

const loadFile = (path: string) =>
  Fs.readFileSync(
    Path.join(__dirname, path),
    'utf8')

const NavBar = loadFile('fixture/Navbar.raw')
const Button = loadFile('fixture/Button.raw')
const Modifiers = loadFile('fixture/Modifiers.raw')

test('get export keys for Navbar', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  expect(fileExportsKeys.length).toBe(8)
})

test('get block from file path', () => {
  const block = getBlockNameFromPath(Path.join(__dirname, '/fixture/Navbar.raw'))
  expect(block).toBe('navbar')
})

test('convert array to union types', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  const sassOptions = getAllStylesFromFile(fileExportsKeys, 'navbar')
  const unions = convertToUnionType(sassOptions)
  expect(unions).toMatchSnapshot()
})

test('combine all variants to one type', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  const sassOptions = getAllStylesFromFile(fileExportsKeys, 'navbar')
  const unions = convertToUnionType(sassOptions)
  const all = combineExistingUnionsToAll(unions)
  expect(all).toMatchSnapshot()
})

test('combine all variants to one type - when some of keys if missing', () => {
  const all = combineExistingUnionsToAll({ modifiers: 'isSth1 | isSth2' })
  expect(all).toMatchSnapshot()
})

test('get all bem styles as well as these which were not recognized', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  const sassOptions = getAllStylesFromFile(fileExportsKeys, 'navbar')
  expect(sassOptions).toMatchSnapshot()
})

test('resolve variants, modifiers and elements from exports', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  const sassOptions = bulmaToBEM(fileExportsKeys, 'navbar')
  expect(sassOptions).toMatchSnapshot()
})

test('export all segregated types as TS union types', () => {
  const fileExportsKeys = getExportKeys(NavBar)
  const sassOptions = bulmaToBEM(fileExportsKeys, 'navbar')
  const exportedTypes = exportBEMTypes(sassOptions)
  expect(exportedTypes).toMatchSnapshot()
})

test('when there is not elements', () => {
  const fileExportsKeys = getExportKeys(Button)
  const sassOptions = bulmaToBEM(fileExportsKeys)
  const exportedTypes = exportBEMTypes(sassOptions)
  expect(exportedTypes).toMatchSnapshot()
})

test('recognize exports when there is a number in name', () => {
  const fileExportsKeys = getExportKeys(Modifiers)
  expect(fileExportsKeys.length).toBe(6)
})