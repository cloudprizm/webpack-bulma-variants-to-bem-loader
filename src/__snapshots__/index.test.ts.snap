// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combine all variants to one type - when some of keys if missing 1`] = `"type All = Modifiers"`;

exports[`combine all variants to one type 1`] = `"type All = Elements | Block | Modifiers | Others"`;

exports[`convert array to union types 1`] = `
Object {
  "block": "type Block = \\"navbar\\"",
  "elements": "type Elements = \\"navbarLink\\" | \\"navbarItem\\" | \\"navbarDropdown\\"",
  "modifiers": "type Modifiers = \\"isArrowless\\" | \\"hasDropdown\\"",
  "others": "type Others = \\"spinAround\\" | \\"dataItem\\"",
}
`;

exports[`export all segregated types as TS union types 1`] = `
Array [
  "export type Elements = \\"navbarLink\\" | \\"navbarItem\\" | \\"navbarDropdown\\"",
  "export type Block = \\"navbar\\"",
  "export type Modifiers = \\"isArrowless\\" | \\"hasDropdown\\"",
  "export type All = Elements | Block | Modifiers",
]
`;

exports[`get all bem styles as well as these which were not recognized 1`] = `
Object {
  "block": Array [
    "navbar",
  ],
  "elements": Array [
    "navbarLink",
    "navbarItem",
    "navbarDropdown",
  ],
  "modifiers": Array [
    "isArrowless",
    "hasDropdown",
  ],
  "others": Array [
    "spinAround",
    "dataItem",
  ],
}
`;

exports[`resolve variants, modifiers and elements from exports 1`] = `
Object {
  "block": Array [
    "navbar",
  ],
  "elements": Array [
    "navbarLink",
    "navbarItem",
    "navbarDropdown",
  ],
  "modifiers": Array [
    "isArrowless",
    "hasDropdown",
  ],
}
`;

exports[`when there is not elements 1`] = `
Array [
  "export type Block = \\"button\\" | \\"icon\\"",
  "export type Modifiers = \\"isLoading\\" | \\"isFocused\\" | \\"isActive\\"",
  "export type All = Block | Modifiers",
]
`;
